x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/settings/AbstractSettingsService.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.relevancesearch.settings;
+
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.index.IndexNotFoundException;
+import org.elasticsearch.xpack.relevancesearch.settings.index.IndexCreationService;
+
+import java.util.Map;
+
+public abstract class AbstractSettingsService<S extends Settings> {
+    public static final String ENT_SEARCH_INDEX = ".ent-search";
+
+    protected final Client client;
+
+    public AbstractSettingsService(final Client client) {
+        this.client = client;
+    }
+
+    protected abstract String getName();
+
+    public S getSettings(String settingsId) throws SettingsNotFoundException, InvalidSettingsException {
+        // TODO cache relevance settings, including cache invalidation
+        Map<String, Object> settingsContent = null;
+        try {
+            settingsContent = client.prepareGet(ENT_SEARCH_INDEX, getSettingsPrefix() + settingsId).get().getSource();
+        } catch (IndexNotFoundException e) {
+            IndexCreationService.ensureInternalIndex(client);
+        }
+
+        if (settingsContent == null) {
+            throw new SettingsNotFoundException(getName() + " settings " + settingsId + " not found");
+        }
+
+        return parseSettings(settingsContent);
+    }
+
+    protected abstract S parseSettings(Map<String, Object> source) throws InvalidSettingsException;
+
+    protected abstract String getSettingsPrefix();
+
+    public static class SettingsNotFoundException extends Exception {
```

Comment:
I see several places where methods `throws` both of these Exceptions. 

Perhaps we could consider a small readability improvement: 
```
public abstract static class SettingsException extends Exception { }
... 
public static class SettingsNotFoundException extends SettingsException {
    public SettingsNotFoundException(String message) { 
        super(message); 
    }
}
```

This would still require the more specific Exceptions to be thrown, but allow us to simplify with `throws SettingsException`. 

What are your thoughts?

Reply:
I like it! I'll use the messages returned from the exceptions to use a single catch. Thanks for the suggestion! 💯